lib/Dialect/Substrait/Transforms/EmitDeduplication.cpp

```diff
@@ -0,0 +1,185 @@
+//===- EmitDeduplication.cpp - Impl. of emit deduplication ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "structured/Dialect/Substrait/Transforms/Passes.h"
+
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "structured/Dialect/Substrait/IR/Substrait.h"
+
+namespace mlir::substrait {
+#define GEN_PASS_DEF_SUBSTRAITEMITDEDUPLICATIONPASS
+#include "structured/Dialect/Substrait/Transforms/Passes.h.inc"
+} // namespace mlir::substrait
+
+using namespace llvm;
+using namespace mlir;
+using namespace mlir::substrait;
+
+namespace {
+
+struct SubstraitEmitDeduplicationPass
+    : substrait::impl::SubstraitEmitDeduplicationPassBase<
+          SubstraitEmitDeduplicationPass> {
+  using substrait::impl::SubstraitEmitDeduplicationPassBase<
+      SubstraitEmitDeduplicationPass>::SubstraitEmitDeduplicationPassBase;
+
+  void runOnOperation() override;
+};
+
+void SubstraitEmitDeduplicationPass::runOnOperation() {
+  mlir::RewritePatternSet patterns(&getContext());
+  populateEmitDeduplicationPatterns(patterns);
+  if (failed(mlir::applyPatternsAndFoldGreedily(getOperation(),
+                                                std::move(patterns)))) {
+    Location loc = getOperation()->getLoc();
+    emitError(loc) << "emit deduplication: pattern application failed";
```

Comment:
I beliece this only fails if it fails to reach convergence after N iterations. Is that sufficient to fail the pass pipeline post?

Reply:
Yes, it is. The function returns `failure` if the pattern application did not converge after the given number of iterations. I am not 100% sure if that's the *only* reason to fail but it is the only one mentioned in the doc and thus seems to be the *typical* reason for failure:

    /// The greedy rewrite may prematurely stop after a maximum number of
    /// iterations, which can be configured in the configuration parameter.
    ...
    /// Returns "success" if the iterative process converged (i.e., fixpoint was
    /// reached) and no more patterns can be matched within the region.